extensions/consensus/consensus.go

```diff
@@ -0,0 +1,240 @@
+// genesis.json, it locates the named hardforks in this registry.
+var Hardforks = map[string]*Hardfork{}
+
+// RegisterHardfork registers a hardfork definition by the name that should be
+// expected in the genesis file. The payload type is registered in the
+// transaction package, but the route is enabled in the tx app at activation.
+func RegisterHardfork(hf *Hardfork) {
+	if _, have := Hardforks[hf.Name]; have {
+		panic(fmt.Sprintf("already have hardfork %q", hf.Name))
+	}
+	Hardforks[hf.Name] = hf
+
+	for _, newPayload := range hf.TxPayloads {
+		transactions.RegisterPayload(newPayload.Type)
+		// NOTE: newPayload.Route stays in the []Hardfork.TxPayloads until
+		// activation, when TxApp/ABCI enable the tx route.
+	}
+}
+
+func HaveHardFork(name string) bool {
+	_, have := Hardforks[name]
+	return have
+}
+
+// Hardfork specifies the fundamental changes affected by a named hardfork. If
+// a field is nil or the zero value for the type, that particular change is not
+// part of the hardfork's definition. When the hardfork is registered, only the
+// defined changes are applied in the relevant parts of kwild.
+type Hardfork struct {
+	// Name is the hard fork's code name.
+	Name string
+	// NOTE: Activation height is specified by genesis.json (or other dynamic
+	// methods like signaling / voting). This struct defines the changes.
+
+	// Optional consensus logic overrides below.  All of these will require
+	// dynamic kwil-db support, such as registering new encoders and transaction
+	// payloads, submitting consensus parameter updates to the consensus engine,
+	// applying state adjustments at activation height in ABCI / TxApp, etc.
+
+	// TxPayloads specifies new transaction payload to recognize at activation.
+	// To modify an existing payload in a backward incompatible way, instead
+	// create a new version of the payload such as PayloadTypeExecuteActionV2.
+	// TODO: consider payload removal/invalidation and replace.
+	TxPayloads []Payload // Type() and tx app Route() implementation
+
+	// ResolutionUpdates are coordinated change to the resolutions extension.
+	// They may be added, modified (redefined), or removed at activation. Any
+	// have the potential to break consensus, and should be done with a Hardfork.
+	ResolutionUpdates []*ResolutionMod
+
+	// Encoders are new encoder types to register *at activation height* for
+	// core/types/serialize.Encode/Decode e.g. Borsch instead of RLP.
+	// EncodingTypeCustom offset should be used as the first possible external
+	// codec's type to leave space for more kwild canonical codecs in the
+	// future. Choose an encoding type that does not collide with other codecs.
+	Encoders []*serialize.Codec // unimplemented!
+
+	// ParamsUpdates are updates to the consensus engine parameters that should
+	// go into affect at after the activation height (returned to the consensus
+	// engine when finalizing the block at this height).
+	ParamsUpdates *ParamUpdates
```

Comment:
> This seems a bit funny, simply because the entire point of consensus params is that they can evolve using pure consensus, rather than a fork, right?

These are the same concepts.  Consensus is the logic that the validators (block producers and approvers) abide by.  Voting is one way to compute the parameter updates fed to cometbft in `FinalizeBlock`.  When you say "pure consensus", you are referring to **governance-based** consensus.  That's a great feature to have, but it's not what defines consensus.  IMO, coordinating the updates to the parameters used by the consensus engine at any given block is how we prevent consensus failure, but as long as it is coordinated, consensus is maintained.  Upgrade by height X or be forced off the network is the most basic form of coordinating rule changes.

The basic ability to seamlessly change rules simply by a consensus among validators (real world, not on-chain governance) deciding to update their genesis configurations with the agreed upon activation height is the most fundamental form of consensus updates.

Reply:
To expand on that, almost any of these rule changes can be done with the voting systems triggering resolutions to put them into affect, but we've focused almost exclusively on federated use cases, which makes this traditional height based activation pretty appropriate.

Also, if you consider doing this through resolutions, the nodes still have to upgrade their software to include the definition of the resolutions, that approach to forward compatibility is extremely limited.

Reply:
I see. Ok, we're on the same page.